### PR TITLE
implement UDAF support

### PR DESCRIPTION
**Pull request by jacksonrnewhouse**
Summary
This PR introduces User-Defined Aggregate Function (UDAF) support to the Arroyo stream processing system. Users can now write Rust UDAFs by defining rust functions with Vec arguments. This includes support for multiple-argument UDF, but does not currently support sorting, due to https://github.com/apache/arrow-datafusion/issues/7531. Once that is fixed we can support it.

Changes
Modified arroyo-sql-macro/src/lib.rs to allow for passing in a UDF block, which can be used for testing UDFs and UDAFs.
Added UDAF support in arroyo-sql/src/expressions.rs.
Introduced a new enum AggregateComputation, which can be either Builtin or UDAF.

Testing
Added new test cases for UDAFs and UDFs.

Backward Compatibility
The changes are backward compatible; existing functionalities should not be affected.

Future Work
Support for aggregate filters and sorts which are currently not supported. Support for two-phase aggregations, which store intermediate computations in the state store.

### CODE REVIEW

File: arroyo-sql/src/lib.rs [hunk 1]
```diff
@@ -180,13 +203,26 @@ impl ArroyoSchemaProvider {
             };
 
             let mut args: Vec<TypeDef> = vec![];
+            let mut vec_arguments = 0;
             for (i, arg) in function.sig.inputs.iter().enumerate() {
+                info!("arg {}: {:?}", i, quote::quote!(#arg).to_string());
```
Comment (mwylde) [inline arroyo-sql/src/lib.rs:208]: this is a bit noisy — maybe downgrade to debug?

File: arroyo-sql/src/lib.rs [hunk 2]
```diff
@@ -180,13 +203,26 @@ impl ArroyoSchemaProvider {
...
                 match arg {
                     FnArg::Receiver(_) => bail!("self types are not allowed in UDFs"),
                     FnArg::Typed(t) => {
-                        args.push((&*t.ty).try_into().map_err(|_| {
-                            anyhow!("Could not convert arg {} into a SQL data type", i)
-                        })?);
+                        if let Some(vec_type) = Self::vec_inner_type(&*t.ty) {
+                            info!("arg {} is a vector", i);
```
Comment (mwylde) [inline arroyo-sql/src/lib.rs:213]: ditto

**Review by mwylde — APPROVED** [review]
Looks great, and love the PR description!

File: arroyo-sql/src/lib.rs (hunk 1 above)
Comment (jacksonrnewhouse) [reply arroyo-sql/src/lib.rs:208]: These were just me trying to understand the existing code, will delete.

**Pull request merged**